File: APP03022-mrch.backend.somx.catalogos-api/src/main/java/com/sodimac/catman/api/mapper/SupplierMapper.java
package com.sodimac.catman.api.mapper;

import java.util.List;

import org.springframework.stereotype.Component;

import com.sodimac.catman.api.model.dto.PaymentConditionDto;
import com.sodimac.catman.api.model.dto.SupplierBlockInfoDto;
import com.sodimac.catman.api.model.dto.SupplierCreateDto;
import com.sodimac.catman.api.model.dto.SupplierDto;
import com.sodimac.catman.api.model.dto.SupplierFilterDto;
import com.sodimac.catman.api.model.dto.SupplierTypeDto;
import com.sodimac.catman.api.model.entity.PaymentCondition;
import com.sodimac.catman.api.model.entity.Supplier;
import com.sodimac.catman.api.model.entity.SupplierBlock;
import com.sodimac.catman.api.model.entity.SupplierType;

@Component
public class SupplierMapper {

    public SupplierDto toDto(Supplier entity) {
        if (entity == null) {
            return null;
        }

        return SupplierDto.builder()
                .id(entity.getId())
                .supplierNumber(entity.getSupplierNumber())
                .rfc(entity.getRfc())
                .businessName(entity.getBusinessName())
                .supplierType(toSupplierTypeDto(entity.getSupplierType()))
                .logo(entity.getLogo())
                .paymentCondition(toPaymentConditionDto(entity.getPaymentCondition()))
                .status(entity.getStatus())
                .build();
    }

    public Supplier toEntity(SupplierCreateDto dto) {
        if (dto == null) {
            return null;
        }

        return Supplier.builder()
                .supplierNumber(dto.getSupplierNumber())
                .rfc(dto.getRfc())
                .businessName(dto.getBusinessName())
                .logo(dto.getLogo())
                .build();
    }

    public List<SupplierDto> toDtoList(List<Supplier> entities) {
        if (entities == null) {
            return null;
        }

        return entities.stream()
                .map(this::toDto)
                .toList();
    }

    public SupplierTypeDto toSupplierTypeDto(SupplierType entity) {
        if (entity == null) {
            return null;
        }

        return SupplierTypeDto.builder()
                .id(entity.getId())
                .code(entity.getCode())
                .description(entity.getDescription())
                .build();
    }

    public PaymentConditionDto toPaymentConditionDto(PaymentCondition entity) {
        if (entity == null) {
            return null;
        }

        return PaymentConditionDto.builder()
                .id(entity.getId())
                .conditionName(entity.getConditionName())
                .days(entity.getDays())
                .build();
    }

    public List<SupplierTypeDto> toSupplierTypeDtoList(List<SupplierType> entities) {
        if (entities == null) {
            return null;
        }

        return entities.stream()
                .map(this::toSupplierTypeDto)
                .toList();
    }

    public List<PaymentConditionDto> toPaymentConditionDtoList(List<PaymentCondition> entities) {
        if (entities == null) {
            return null;
        }

        return entities.stream()
                .map(this::toPaymentConditionDto)
                .toList();
    }

    /**
     * STM-831: Convierte un Supplier a SupplierFilterDto con informacion de bloqueo.
     */
    public SupplierFilterDto toFilterDto(Supplier entity, boolean blocked, SupplierBlock activeBlock) {
        if (entity == null) {
            return null;
        }

        SupplierFilterDto.SupplierFilterDtoBuilder builder = SupplierFilterDto.builder()
                .id(entity.getId())
                .supplierNumber(entity.getSupplierNumber())
                .rfc(entity.getRfc())
                .businessName(entity.getBusinessName())
                .supplierType(toSupplierTypeDto(entity.getSupplierType()))
                .paymentCondition(toPaymentConditionDto(entity.getPaymentCondition()))
                .status(entity.getStatus())
                .blocked(blocked);

        if (blocked && activeBlock != null) {
            builder.blockInfo(toBlockInfoDto(activeBlock));
        }

        return builder.build();
    }

    /**
     * STM-831: Convierte un SupplierBlock a SupplierBlockInfoDto.
     */
    public SupplierBlockInfoDto toBlockInfoDto(SupplierBlock entity) {
        if (entity == null) {
            return null;
        }

        return SupplierBlockInfoDto.builder()
                .validFrom(entity.getValidFrom())
                .validTo(entity.getValidTo())
                .blockReason(entity.getBlockReason())
                .build();
    }
}

File: APP03022-mrch.backend.somx.util-api/src/mappers/supplier.mapper.ts
import type { Supplier } from '@/entities/Supplier.entity.js';
import type { SupplierType } from '@/entities/SupplierType.entity.js';
import type { PaymentCondition } from '@/entities/PaymentCondition.entity.js';
import type { SupplierBlock } from '@/entities/SupplierBlock.entity.js';
import type {
    SupplierDto,
    SupplierTypeDto,
    PaymentConditionDto,
    SupplierFilterDto,
    SupplierBlockInfoDto,
    SupplierCreateDto
} from '@/dto/supplier.dto.js';

export function toSupplierTypeDto(entity: SupplierType | null | undefined): SupplierTypeDto | null {
    if (!entity) return null;
    return {
        id: entity.id,
        code: entity.code,
        description: entity.description
    };
}

export function toPaymentConditionDto(entity: PaymentCondition | null | undefined): PaymentConditionDto | null {
    if (!entity) return null;
    return {
        id: entity.id,
        conditionName: entity.conditionName,
        days: entity.days
    };
}

export function toDto(entity: Supplier | null | undefined): SupplierDto | null {
    if (!entity) return null;
    return {
        id: entity.id,
        supplierNumber: entity.supplierNumber,
        rfc: entity.rfc,
        businessName: entity.businessName,
        supplierType: toSupplierTypeDto(entity.supplierType),
        logo: entity.logo ?? null,
        paymentCondition: toPaymentConditionDto(entity.paymentCondition),
        emailPrincipal: entity.emailPrincipal ?? null,
        emailFinancial: entity.emailFinancial ?? null,
        emailCommercial: entity.emailCommercial ?? null,
        status: entity.status
    };
}

export function toDtoList(entities: Supplier[] | null | undefined): SupplierDto[] {
    if (!entities) return [];
    return entities.map(e => toDto(e)).filter((d): d is SupplierDto => d !== null);
}

export function toSupplierTypeDtoList(entities: SupplierType[] | null | undefined): SupplierTypeDto[] {
    if (!entities) return [];
    return entities.map(e => toSupplierTypeDto(e)).filter((d): d is SupplierTypeDto => d !== null);
}

export function toPaymentConditionDtoList(entities: PaymentCondition[] | null | undefined): PaymentConditionDto[] {
    if (!entities) return [];
    return entities.map(e => toPaymentConditionDto(e)).filter((d): d is PaymentConditionDto => d !== null);
}

export function toEntity(dto: SupplierCreateDto): Partial<Supplier> {
    return {
        supplierNumber: dto.supplierNumber,
        rfc: dto.rfc,
        businessName: dto.businessName,
        logo: dto.logo ?? null,
        emailPrincipal: dto.emailPrincipal,
        emailFinancial: dto.emailFinancial,
        emailCommercial: dto.emailCommercial ?? null
    };
}

export function toBlockInfoDto(entity: SupplierBlock | null | undefined): SupplierBlockInfoDto | null {
    if (!entity) return null;
    return {
        validFrom: entity.validFrom,
        validTo: entity.validTo,
        blockReason: entity.blockReason ?? null
    };
}

export function toFilterDto(
    entity: Supplier,
    blocked: boolean,
    activeBlock: SupplierBlock | null
): SupplierFilterDto {
    const dto: SupplierFilterDto = {
        id: entity.id,
        supplierNumber: entity.supplierNumber,
        rfc: entity.rfc,
        businessName: entity.businessName,
        supplierType: toSupplierTypeDto(entity.supplierType),
        paymentCondition: toPaymentConditionDto(entity.paymentCondition),
        status: entity.status,
        blocked,
        blockInfo: null
    };
    if (blocked && activeBlock) {
        dto.blockInfo = toBlockInfoDto(activeBlock);
    }
    return dto;
}

